### PR TITLE
buffer: add copyOut() method

### PR DESCRIPTION
**Pull request by mattklein123**
Signed-off-by: Matt Klein <mklein@lyft.com>

### CODE REVIEW

File: include/envoy/buffer/buffer.h [hunk 1]
```diff
@@ -53,6 +53,14 @@ class Instance {
   virtual void commit(RawSlice* iovecs, uint64_t num_iovecs) PURE;
 
   /**
+   * Copy out a section of the buffer.
+   * @param data supplies the output buffer to fill.
+   * @param size supplies the size of the output buffer.
+   * @param start supplies the buffer index to start copying from.
+   */
+  virtual void copyOut(void* data, uint64_t size, size_t start) const PURE;
```
Comment (htuch) [inline include/envoy/buffer/buffer.h]: Maybe order params (start, size, data), I don't know if we formalize it, but it's kind of nice to have IN, IN/OUT, OUT ordering.

File: source/common/buffer/buffer_impl.cc [hunk 2]
```diff
@@ -41,6 +41,19 @@ void OwnedImpl::commit(RawSlice* iovecs, uint64_t num_iovecs) {
   UNREFERENCED_PARAMETER(rc);
 }
 
+void OwnedImpl::copyOut(void* data, uint64_t size, size_t start) const {
+  ASSERT(start + size <= length());
+
+  evbuffer_ptr start_ptr;
+  int rc = evbuffer_ptr_set(buffer_.get(), &start_ptr, start, EVBUFFER_PTR_SET);
+  ASSERT(rc != -1);
```
Comment (htuch) [inline source/common/buffer/buffer_impl.cc]: I guess I don't understand the failure mode for this function, but seem OK?

File: source/common/buffer/buffer_impl.cc [hunk 3]
```diff
@@ -41,6 +41,19 @@ void OwnedImpl::commit(RawSlice* iovecs, uint64_t num_iovecs) {
...
+  UNREFERENCED_PARAMETER(rc);
+
+  ev_ssize_t copied = evbuffer_copyout_from(buffer_.get(), &start_ptr, data, size);
```
Comment (htuch) [inline source/common/buffer/buffer_impl.cc]: Again, it's a bit black magic when we can fail vs. not fail, the docs don't really say much.

File: test/common/buffer/watermark_buffer_test.cc [hunk 4]
```diff
@@ -18,8 +20,19 @@ class WatermarkBufferTest : public testing::Test {
...
+
 TEST_F(WatermarkBufferTest, TestWatermark) { ASSERT_EQ(10, buffer_.highWatermark()); }
 
+TEST_F(WatermarkBufferTest, CopyOut) {
+  buffer_.add("hello world");
+  std::array<char, 5> out;
+  buffer_.copyOut(out.data(), out.size(), 0);
+  EXPECT_EQ(std::string(out.data(), out.size()), "hello");
+
+  buffer_.copyOut(out.data(), out.size(), 6);
```
Comment (htuch) [inline test/common/buffer/watermark_buffer_test.cc]: What about copying 0 size?

File: include/envoy/buffer/buffer.h (hunk 1 above)
Comment (mattklein123) [reply include/envoy/buffer/buffer.h]: I don't really care one way or the other, but we don't do this anywhere else (even in this file), so it seems strange to do it here. But will change.

File: source/common/buffer/buffer_impl.cc (hunk 2 above)
Comment (mattklein123) [reply source/common/buffer/buffer_impl.cc]: Can't fail with proper input params which we verify.

File: source/common/buffer/buffer_impl.cc (hunk 3 above)
Comment (mattklein123) [reply source/common/buffer/buffer_impl.cc]: Can't fail with proper input params which we verify.

File: test/common/buffer/watermark_buffer_test.cc (hunk 4 above)
Comment (mattklein123) [reply test/common/buffer/watermark_buffer_test.cc]: will add

**Comment by mattklein123** [comment]
@htuch updated

**Pull request merged**